source/common/stats/thread_local_store.cc

```diff
@@ -54,6 +56,29 @@ std::list<GaugeSharedPtr> ThreadLocalStoreImpl::gauges() const {
     for (auto gauge : scope->central_cache_.gauges_) {
       if (names.insert(gauge.first).second) {
         ret.push_back(gauge.second);
+      } else {
+        ENVOY_LOG(warn, "duplicate guage {}.{}", scope->prefix_, gauge.first);
+      }
+    }
+  }
+
+  return ret;
+}
+
+std::list<HistogramSharedPtr> ThreadLocalStoreImpl::histograms() const {
+  // Handle de-dup due to overlapping scopes.
+  std::list<HistogramSharedPtr> ret;
+  std::unordered_set<std::string> names;
+  std::unique_lock<std::mutex> lock(lock_);
+  // TODO(ramaraochavali): incorporate the scopes into the histogram names.
+  for (ScopeImpl* scope : scopes_) {
+    for (auto name_histogram_pair : scope->central_cache_.histograms_) {
+      const std::string& hist_name = name_histogram_pair.first;
+      const ParentHistogramSharedPtr& parent_hist = name_histogram_pair.second;
+      if (names.insert(hist_name).second) {
+        ret.push_back(parent_hist);
+      } else {
+        ENVOY_LOG(warn, "duplicate histogram {}.{}", scope->prefix_, hist_name);
```

Comment:
OK we're still in the situation where I don't think this should really happen.  But it does during hot_restart tests, and we should  at least leave a TODO and figure out why.

And I'd like to know if this happens for histograms only, or for all the stats.

Reply:
It can happen per above, please remove. If you are interested to see which are duplicated, you can reduce all of these log messages to debug to see which ones are overlaid (feel free to print out the log messages during the hot restart test and paste them here in case we want to check...)

Reply:
I have removed for others and changed histogram one to debug

Reply:
@mattklein123 per @ramaraochavali's earlier comments, when he had this as an assert, it fired only during hot-restart tests, which makes me wonder whether it's really a scope issue or an issue about the semantics of state resurrection during hot restart in particular.

And apologies to @ramaraochavali and @mattklein123 about the review-churn here but it wasn't obvious from looking at this code-change in isolation how it's supposed to work.  If we can wind up at least with more comments about the semantics here it'd be useful for the future.

Reply:
@jmarantz do you want any additional comments here? if yes, can you let me know what you would like to add?